pylint/checkers/variables.py

```diff
@@ -979,267 +1003,298 @@ def visit_assignname(self, node: nodes.AssignName) -> None:
-                        #      f = 42
-                        if (
-                            isinstance(frame, nodes.ClassDef)
-                            and node.name in frame.locals
-                        ):
-                            if isinstance(node.parent, nodes.Arguments):
-                                if stmt.fromlineno <= defstmt.fromlineno:
-                                    # Doing the following is fine:
-                                    #   class A:
-                                    #      x = 42
-                                    #      y = lambda attr=x: attr
-                                    self.add_message(
-                                        "used-before-assignment",
-                                        args=node.name,
-                                        node=node,
-                                    )
-                            else:
-                                self.add_message(
-                                    "undefined-variable", args=node.name, node=node
-                                )
-                        elif current_consumer.scope_type == "lambda":
-                            self.add_message(
-                                "undefined-variable", node=node, args=node.name
-                            )
-                elif self._is_only_type_assignment(node, defstmt):
-                    self.add_message("undefined-variable", node=node, args=node.name)
-                elif isinstance(defstmt, nodes.ClassDef):
-                    is_first_level_ref = self._is_first_level_self_reference(
-                        node, defstmt
-                    )
-                    if is_first_level_ref == 2:
                         self.add_message(
-                            "used-before-assignment", node=node, args=node.name
+                            "undefined-variable", args=node.name, node=node
                         )
-                    if is_first_level_ref:
-                        break
-                elif isinstance(defnode, nodes.NamedExpr):
-                    if isinstance(defnode.parent, nodes.IfExp):
-                        if self._is_never_evaluated(defnode, defnode.parent):
+                        return (VariableVisitConsumerAction.CONSUME, found_nodes)
+
+            elif base_scope_type != "lambda":
+                # E0601 may *not* occurs in lambda scope.
+
+                # Handle postponed evaluation of annotations
+                if not (
+                    self._postponed_evaluation_enabled
+                    and isinstance(stmt, (nodes.AnnAssign, nodes.FunctionDef))
+                ):
+                    self.add_message(
+                        "used-before-assignment", args=node.name, node=node
+                    )
+                    return (VariableVisitConsumerAction.CONSUME, found_nodes)
+
+                # E0601 can occur in class-level scope in lambdas, as in
```

Comment:
```suggestion

            elif base_scope_type == "lambda":
                # E0601 can occur in class-level scope in lambdas, as in
```

Reply:
I now know why this wrong in the first place. This `elif` doesn't make a lot sense (I think?) 
https://github.com/PyCQA/pylint/blob/6caab13816e8098063b6e077ab0044b94aa87557/pylint/checkers/variables.py#L1165-L1176

Reply:
That's why I removed the second elif, but then the indentation was messed up.

Reply:
I think the latest commit fixes this (https://github.com/PyCQA/pylint/pull/5241/commits/5c99a866fc6174f49d0e06105f6dc5037c4aebc2)

Now there is one `elif base_scope_type != "lambda":` and then the `elif`'s that were previously under `elif base_scope_type == "lambda": `.
Using `else` doesn't work because then a `else-if-used` warning is raised.

Reply:
That still missed the `base_scope_type == "lambda"` condition. So "luck" if it doesn't break.
Using `else` would be wrong, as there is another `elif` case afterwards.

Reply:
But if `elif base_scope_type != "lambda":` is `False`  (L1165) isn't `base_scope_type == "lambda":` then always `True`?

Reply:
Hmm, true. However, adding a separate `elif` makes that more explicit. Especially with the block indentation. I just got confused about it in my last comment when I didn't realize that `elif current_consumer.scope_type == "lambda"` isn't an `elif` on the outer if, instead only for the `== "lambda" case.

Reply:
Yeah, looking back at the original version the real mistake was `if isinstance(frame, nodes.ClassDef) ...` instead of `elif isinstance(frame, nodes.ClassDef) ...`

I'm not sure what is better and don't really have a preference. For me the current version works (and has less indentation), but I'm fine with adding the unnecessary `if` back agin. What do you want?

Reply:
I would prefer to stay as close to the original as possible. So with the `elif base_scope_type == "lambda"`

Reply:
Done!